### PR TITLE
feat: support export data

### PR DESCRIPTION
**Pull request by rumor-sourse**
### What problem does this PR solve?

Support export data to txt file , csv file or remote openGemini server.

### What is changed and how it works?
You need to create some data in openGemini.for example NOAA_water_database.
```shell
curl -G  https://s3.amazonaws.com/noaa.water-database/NOAA_data.txt > NOAA_data.txt
ts-cli import --path=NOAA_data.txt --host=127.0.0.1 --port=8086 --precision=s
```
1. Export data to txt file or csv file
```shell
ts-cli export --format txt --out /tmp/openGemini/export/export.txt --data /tmp/openGemini/data --wal /tmp/openGemini/data --dbfilter NOAA_water_database
ts-cli export --format csv --out /tmp/openGemini/export/export.csv --data /tmp/openGemini/data --wal /tmp/openGemini/data --dbfilter NOAA_water_database
```
2. Export data to txt file or csv file using mstfilter, timefilter or retentionfilter
```shell
ts-cli export --format txt --out /tmp/openGemini/export/export.txt --data /tmp/openGemini/data --wal /tmp/openGemini/data 
--dbfilter NOAA_water_database --mstfilter h2o_pH --timefilter "2019-08-25T09:18:00Z~2019-08-26T07:48:00Z"

ts-cli export --format csv --out /tmp/openGemini/export/export.csv --data /tmp/openGemini/data --wal /tmp/openGemini/data 
--dbfilter NOAA_water_database --mstfilter h2o_pH --timefilter "2019-08-25T09:18:00Z~2019-08-26T07:48:00Z"
```
3. Export data to remote openGemini
```shell
ts-cli export --format remote --remote ${host}:8086 --data /tmp/openGemini/data --wal /tmp/openGemini/data  
--dbfilter NOAA_water_database --mstfilter h2o_feet
```

### How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] TestDatabaseDiskInfo_init
- [x] TestExporter_ExportTxt
- [x] TestExporter_ExportCsv
- [x] TestExporter_writeDDL
- [x] TestExporter_writeDML

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by xuthus5** [comment]
please rebase your commit.

File: app/ts-cli/cmd/export.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
+package cmd
```
Comment (xuthus5) [inline app/ts-cli/cmd/export.go:1]: The LICENSE description needs to be added to the file header.

File: app/ts-cli/cmd/export.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+import (
+	"bufio"
+	"encoding/json"
+	"flag"
+	"github.com/openGemini/openGemini/app/ts-cli/geminicli"
```
Comment (xuthus5) [inline app/ts-cli/cmd/export.go:7]: Use `goimports-reviser` to sort packages, such as: `goimports-reviser -project-name none-pjn ./...`

File: app/ts-cli/cmd/export.go [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+			if err := exportCmd.Export(&options, nil); err != nil {
+				return err
+			}
+		}
+		return nil
+	},
+}
+
+func getResumeConfig() (*geminicli.CommandLineConfig, error) {
+	jsonFile, err := os.Open(geminicli.ResumeJsonPath)
```
Comment (xuthus5) [inline app/ts-cli/cmd/export.go:89]: Replace `os.Open` with `os.ReadFile` is a better way to read only the file content.

File: app/ts-cli/geminicli/export.go [hunk 4]
```diff
@@ -0,0 +1,1786 @@
...
+	var start, end string
+	timeSlot := strings.Split(clc.TimeFilter, "~")
+	if len(timeSlot) == 2 {
+		start = timeSlot[0]
+		end = timeSlot[1]
+	} else if clc.TimeFilter != "" {
+		return fmt.Errorf("invalid time filter %q", clc.TimeFilter)
+	}
+	// set defaults
+	if start != "" {
```
Comment (xuthus5) [inline app/ts-cli/geminicli/export.go:136]: Replace if-else with default assignment.

**Comment by xuthus5** [comment]
Replace title `Feat` to `feat`.

File: app/ts-cli/geminicli/export.go [hunk 5]
```diff
@@ -0,0 +1,1786 @@
...
+		if !info.IsDir() || path == baseDir {
+			return nil
+		}
+		dirs = append(dirs, path)
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	sort.Slice(dirs, func(i, j int) bool {
```
Comment (xuthus5) [inline app/ts-cli/geminicli/export.go:98]: `Sort.Strings` may be more efficient if you only want to take the first or last element of a sorted string slice.

File: app/ts-cli/geminicli/export.go [hunk 6]
```diff
@@ -0,0 +1,1786 @@
...
+		rpToWalDirMap:   make(map[string]string),
+		rpToIndexDirMap: make(map[string]string),
+	}
+}
+
+func (d *DatabaseDiskInfo) init(actualDataDir string, actualWalDir string, databaseName string, retentionPolicy string) error {
+	d.dbName = databaseName
+
+	// check whether the database is in actualDataPath
+	dataDir := path.Join(actualDataDir, databaseName)
```
Comment (xuthus5) [inline app/ts-cli/geminicli/export.go:225]: `path.Join` use unix style slash `/` to combie string, on windows platform may cause problem.

File: app/ts-cli/geminicli/export.go [hunk 7]
```diff
@@ -0,0 +1,1786 @@
...
+	}
+	for _, ptDir := range ptDirs {
+		// ie. /tmp/openGemini/data/data/my_db/0
+		ptTsspPath := path.Join(d.dataDir, ptDir.Name())
+		// ie. /tmp/openGemini/data/wal/my_db/0
+		ptWalPath := path.Join(d.walDir, ptDir.Name())
+
+		if retentionPolicy != "" {
+			ptWithRp := ptDir.Name() + ":" + retentionPolicy
+			rpTsspPath := path.Join(ptTsspPath, retentionPolicy)
```
Comment (xuthus5) [inline app/ts-cli/geminicli/export.go:250]: We need to examine the path splicing of different OS.

File: app/ts-cli/geminicli/export.go [hunk 8]
```diff
@@ -0,0 +1,1786 @@
...
+			} else {
+				d.rps[ptWithRp] = struct{}{}
+				d.rpToTsspDirMap[ptWithRp] = rpTsspPath
+				d.rpToIndexDirMap[ptWithRp] = path.Join(rpTsspPath, "index")
+			}
+			rpWalPath := path.Join(ptWalPath, retentionPolicy)
+			if _, err := os.Stat(rpWalPath); err != nil {
+				return fmt.Errorf("retention policy %q invalid : %s", retentionPolicy, err)
+			} else {
+				d.rpToWalDirMap[ptWithRp] = rpWalPath
```
Comment (xuthus5) [inline app/ts-cli/geminicli/export.go:262]: If `os.Stat` result is nil, `d.rpToWalDirMap` will not set a value. Is this expected?

File: app/ts-cli/geminicli/export.go [hunk 9]
```diff
@@ -0,0 +1,1786 @@
...
+			}
+			continue
+		}
+
+		rpTsspDirs, err1 := os.ReadDir(ptTsspPath)
+		if err1 != nil {
+			return err1
+		}
+		for _, rpDir := range rpTsspDirs {
+			if rpDir.IsDir() {
```
Comment (xuthus5) [inline app/ts-cli/geminicli/export.go:272]: Use the guard statement to reduce the if judgment.

**Comment by xuthus5** [comment]
We'd better describe how this feature works.

File: app/ts-cli/cmd/export.go [hunk 10]
```diff
@@ -0,0 +1,135 @@
+// Copyright 2022 Huawei Cloud Computing Technologies Co., Ltd.
```
Comment (xuthus5) [inline app/ts-cli/cmd/export.go:1]: replace copyright year to 2024, you can change author to `openGemini author` like this: `Copyright 2024 The openGemini Authors All rights reserved.`.

**Comment by xuthus5** [comment]
If you have resolved any review change, please mark it.

**Comment by xiangyu5632** [comment]
Why we need to upgrade go version to go v1.21？

File: app/ts-cli/cmd/export.go [hunk 11]
```diff
@@ -0,0 +1,136 @@
+// Copyright 2024 Huawei Cloud Computing Technologies Co., Ltd.
```
Comment (xiangyu5632) [inline app/ts-cli/cmd/export.go:1]: openGemini authors

**Comment by xiangyu5632** [comment]
> TL;DR please rebase your commit and signed off change. `git commit -s -m "your commit message"`

File: app/ts-cli/geminicli/export.go [hunk 12]
```diff
@@ -0,0 +1,1804 @@
...
+		return t.UnixNano(), nil
+	}
+
+	// convert timeStamp
+	timestamp, err := strconv.ParseInt(input, 10, 64)
+	if err == nil {
+		return timestamp, nil
+	}
+
+	// err
```
Comment (shilinlee) [inline app/ts-cli/geminicli/export.go:1802]: remove the comment

**Review by shilinlee — CHANGES_REQUESTED** [review]

File: app/ts-cli/tests/export/init.go [hunk 13]
```diff
@@ -0,0 +1,448 @@
...
+	for i := range rs {
+		sort.Sort(&rs[i].Fields)
+	}
+	err = sh.WriteRows(rs, buff)
+	if err != nil {
+		return err
+	}
+
+	// wait index flush
+	//time.Sleep(time.Second * 1)
```
Comment (shilinlee) [inline app/ts-cli/tests/export/init.go:259]: remove unused code

**Review by shilinlee — CHANGES_REQUESTED** [review]

**Comment by xiangyu5632** [comment]
Please re-describe the content of the PR

**Comment by xuthus5** [comment]
> 放一个tests目录在 app/ts-cli 下面中感觉怪怪的。如果是功能测试用例，我建议放到AT用例的tests目录下。 cc @shilinlee @xuthus5 @fx408 ，你们觉得呢？

I think it's the better way.

**Comment by xuthus5** [comment]
The evolution strategy of ts-cli will be discussed at the community meeting in October, and this submission is temporarily suspended.

**Comment by shilinlee** [comment]
> The evolution strategy of `ts-cli` will be discussed at the community meeting in October, and this submission is temporarily suspended.↳

This feature has been fully tested and there is no problem in merging it into the feature branch.

**Review by shilinlee — APPROVED** [review]
LGTM
